Name - Shreshth Sinha


Student ID - 8895491

Loading required libraries

In [1]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error, r2_score
import numpy as np
import pandas as pd

Loading the dataset

In [2]:
df_diabetes = load_diabetes()
X, y = df_diabetes.data, df_diabetes.target

Spliting Data set into training and testing

In [3]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Perform Cross validation

In [5]:
degrees = range(9) 
results = []

for degree in degrees:
    md = make_pipeline(PolynomialFeatures(degree), LinearRegression())
    md.fit(X_train, y_train)
    y_pred = md.predict(X_test)
    
    # We will calculate the r2 score
    r2 = r2_score(y_test, y_pred)
    
    # We will calculate the mean absolute error score
    mae = mean_absolute_error(y_test, y_pred)
    
    # We will calculate the mean absolute percentage error score
    mape = mean_absolute_percentage_error(y_test, y_pred) 
    
    results.append((degree, r2, mae, mape))

Storing results

In [6]:
df_res = pd.DataFrame(results, columns=['Degree', 'R-Squared', 'MAE', 'MAPE'])

Calculating mean and standard deviation

In [7]:
mn_std_df = df_res.groupby('Degree').agg({
    'R-Squared': ['mean', 'std'],
    'MAE': ['mean', 'std'],
    'MAPE': ['mean', 'std']
})

# We will display the summary table
print("Summary Table:\n",mn_std_df)

Summary Table:
         R-Squared             MAE          MAPE    
             mean std        mean std      mean std
Degree                                             
0       -0.011963 NaN   64.006461 NaN  0.627918 NaN
1        0.452603 NaN   42.794095 NaN  0.374998 NaN
2        0.415640 NaN   43.581693 NaN  0.382857 NaN
3      -15.627261 NaN  179.775281 NaN  1.637362 NaN
4      -26.728083 NaN  261.667144 NaN  2.300991 NaN
5      -25.992920 NaN  255.968358 NaN  2.270202 NaN
6      -25.975743 NaN  255.908618 NaN  2.269658 NaN
7      -25.975503 NaN  255.906927 NaN  2.269650 NaN
8      -25.975482 NaN  255.906853 NaN  2.269649 NaN


The summary table indicates that there is only one data point (NaN) for each degree that can be evaluated. This may be the result of certain dataset characteristics or may be based on how the data is split during the cross-validation process. For some particular degrees, the data may be extremely scarce. The following are possible causes:
1.The best model recognition: R-square, MAE, and MAPE measures allow us to determine which model will exhibit the highest performance. We can see that the model with degree 1 (quadratic polynomial) has the highest R-squared (0.45) and the lowest MAE (42.79) and MAPE (0.37) based on the mean R-Squared, MAE, and MAPE metrics. As a result, polynomial models can be seen performing well

Why we selected that particular model (Degree 1):
R-Squared: This degree 1 model has the greatest R-Squared value, which shows that it explains a considerable amount of the data's variance. This indicates that the quadratic model fits the data significantly more effectively.

Mean Absolute Error (MAE) and Mean Absolute Percentage Error (MAPE): The model with degree 1 also has the lowest MAE and MAPE. This shows that, in terms of predictive accuracy, it has the smallest average absolute difference between projected and actual values.
 
2.Additional evaluation and interpretation of the performance of the models:  
If an analysis can offer at least one useful insight regarding the selection of the best model, or about features of the chosen one (for example, taking into account an analysis of when it is failing), then the insights for the necessary metrics can be investigated further. It is also crucial to notice that the R-Squared values are perceived as notably negative for higher degree models (for example, degrees 3 to 8). This shows how poorly these models are performing and suggests that they may have overfitted to the training set. This is a result of the models, which develop into very sophisticated representations of the data noise.

As a result, the quadratic model (degree 1) is advised for use in making predictions since it exhibits a good balance between model complexity and predictive ability. When selecting a final model, it is necessary to take into account additional aspects like interpretability and computational resources. Regularised regression is one of the more sophisticated approaches that can be useful for additional investigation.
